[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104215
* #103910
* __->__ #103235

# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: This subclass of `CudaRNGStateTracker` defines the default policy of how RNG states should be shared and synchronized among all ranks to respect the semantics of DTensor random operators.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

XilunWu committed:
[DTensor][Random] Introduce CudaRNGStatesTracker to maintain the default and model parallel seeds

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStatesTracker to maintain the default and model parallel seeds"




[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/rng_tracker.py:
```diff
@@ -0,0 +1,37 @@
+from torch import Tensor
+
+
+class CudaRNGStatesTracker:
```
meta comment: I think this class and `ParallelRNGStyle` should be merged, and we should put this and the global variable in `random.py` instead

XilunWu replied:
The main reason I separate them is to remove circular import dependency:

In DeviceMesh, we use RNG tracker to synchronize. But some logic in Random.py uses DeviceMesh and this causes circular dependency. 

Btw I realize that MegatronStyle can be moved to TP folder.

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        default_seed: int = 1234,
+        tp_dim: int = 0,
+    ):
+        # Megatron maintains two sets of RNG states:
```
this comment should be in the class level with """ """ as comment block

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        # set offset before entering the parallel region
+        mesh = spec.mesh
+        old_offset = _get_rng_offset(mesh)
+        set_pre_op_offset(spec)
```
I think we should add all those methods in this file to this class, and rename this to `DTensorRNGTracker` instead.

XilunWu replied:
I can move them into DtensorRNG. But merging tracker and parallel style caused circular dependency.

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+    # the current rank is in mesh
+    if device_mesh.get_coordinate() is not None:
+        if isinstance(_parallel_rng_style, MegatronRNG):
+            _parallel_rng_style._manual_seed(device_mesh, seed, tp_dim)
```
we need tp_dim for megatronRNG only?

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        `manual_seed` will not set its GPU device's generator seed.
+        Current implementation only supports a GPU device mesh.
+    """
+    # allgather the seed from rank 0 over the default PG
```
this is not from rank 0 right?

wanchaol reviewed (commented):
first pass, see inline comments

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        # calculate tensor_model_parallel_seed as Megatron does
+        # source: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py
+        tensor_model_parallel_rank = device_mesh.get_coordinate()[tp_dim]
+        tensor_model_parallel_seed = default_seed + 2718 + tensor_model_parallel_rank
```
I think this `2718` is a magic number here and if you want to reuse it can we make it a const instead of hard-code it in the code?

XilunWu replied:
It can be any positive number so the value itself won't make any impact on correctness only if this value is the same among all ranks but this is SPMD so no problem. That's why I don't think this value is "magical" enough to be made a constant variable.

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        tensor_model_parallel_rank = device_mesh.get_coordinate()[tp_dim]
+        tensor_model_parallel_seed = default_seed + 2718 + tensor_model_parallel_rank
+
+        # store tensor model parallel RNG state
+        torch.cuda.manual_seed(tensor_model_parallel_seed)
```
ditto.

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+        tp_dim: int = 0,
+    ):
+        # Megatron maintains two sets of RNG states:
+        #   1. default state: torch's cuda RNG state. This is set to the same
+        #   across all ranks unless `data_parallel_random_init` is explicitly
+        #   set to True. Megatron uses this RNG state for non-tensor-parallel
+        #   regions.
+        #   2. tensor model parallel state: separate from the default state
+        #   and is only used in tensor-parallel regions. This is set to the same
+        #   across tensor parallel groups (i.e. along non-tp dimension) but different
+        #   within a tensor parallel group.
```
What is the UX for 2D parallel? Can you provide a user flow here? Because in 2D parallel, you need to differentiate whether it's TP or SP and whether it's within TP or not, right? How are going to handle that?

XilunWu replied:
basically to support SP we only need to add a `DistributedRNGStyle` that implements the corresponding `distribute_region` context and call it within SP's API:

```
def SomeSPApi():
    with SequenceParallelRNGStyle.distribute_region():
        # perform SP computation here
```

fduwjj replied:
Well, if for initialization, user can do this one-time thing, but we don't want to call cx within forward code right?

fduwjj reviewed (commented):
Thanks for working on this. I leave some comments and questions here.

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+
+
+# TODO: change this to ABC
+class ParallelRNGStyle:
+    def __init__(self):
+        pass
```
Do I understand correctly that you want to do something like this:
```
# non-TP style for initialization
_parallel_rng_style = DTensorParallelRNG()
dtensor.uniform_()

# use TP style for random dropout
_parallel_rng_style = TensorParallelRNG()
dtensor.dropout()
```

We offer this flexibility by assigning different values to global `_parallel_rng_style`.

fduwjj replied:
Do we want to explicitly expose this to users? I thought this will just be one entry of `DeviceMesh`?

XilunWu replied:
> Do we want to explicitly expose this to users? I thought this will just be one entry of `DeviceMesh`?

No this shouldn't be facing to users directly but we can have this within TP's API implementation. I'm still thinking about where to put this variable. Currently it is a module-level variable within `random.py` but maybe we can put it inside device mesh @wanchaol WDYT?

XilunWu replied:
updated the code snippet. Getting rid of the `_parallel_rng_style` variable seems to be simpler.

wanchaol replied:
I think it should just be the module level global variable. Inside TP's API it can override this to take TPRNG.

Also, I think we should NOT make this a DeviceMesh attribute as it's already a global attribute, we should make sure RNGTrackers are lazily initialized so that we could put those initialization logic in different APIs (i.e. `distribute_tensor` or `parallelize_module`)

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+            _rng_tracker.set_rng_state("parallel-rng", torch.cuda.get_rng_state())
+
+    @contextlib.contextmanager
+    def _parallel_region(self, spec: DTensorSpec):
```
sounds good!

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -301,3 +261,158 @@ def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
+            torch.cuda.set_rng_state(old_state)
+
+
+class MegatronRNG(ParallelRNGStyle):
```
Yes. I originally called it because I want to match Megatron's behavior. TPRNG seems a better name.

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStatesTracker to maintain the default and model parallel seeds"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStatesTracker`: A `CudaRNGStatesTracker` object named `_rng_tracker` is responsible for

- store RNG state tensor in a dict
- synchronize RNG state tensor if it has not been stored in the dict, by broadcasting the current RNG state from rank 0

2. `ParallelRNGStyle`: A base class that contains two interfaces

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
2. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: 

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
4. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: 

-------WIP-------

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
4. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +59,242 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+        if name not in self.rng_states:
+            with torch.random.fork_rng(self._devices):
+                torch.cuda.manual_seed(seed)
+                self.rng_states[name] = torch.cuda.get_rng_state()
```
as discussed offline, we can probably simplify this without fork_rng

XilunWu replied:
done

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: 

-------WIP-------

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
4. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +59,242 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
 
-def _get_rng_offset(device_mesh: DeviceMesh) -> int:
-    """Returns the random number generator state offset for the calling rank.
+    def reset(self) -> None:
```
it looks like this `reset` does not get used anywhere, shall we delete it?

XilunWu replied:
True. I'll remove it.

wanchaol reviewed (commented):
Please address comments as we discussed offline.

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +53,239 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+    def __init__(self):
+        super().__init__()
+        # synchronize random seed to initialize RNG state
+        _seed = [torch.cuda.initial_seed()]
```
That's also a good way to synchronize. I don't mind using that.

XilunWu commented on torch/distributed/_tensor/api.py:
```diff
@@ -396,6 +401,14 @@ def distribute_tensor(
+
+    # instantiate a RNG tracker if haven't. By default DTensor uses an
+    # OffsetBasedRNGTracker to perform random operators.
+    # TODO: the value assignment to global variable is not the ideal solution
```
One alternative is we instantiate `_rng_tracker = OffsetBasedRNGTracker()` and add a flag attribute `is_sync = False` to it.

fduwjj replied:
Do we want to have an option for power user to swap between different RNG Tracker?

XilunWu replied:
I think what we can do is to add an extra argument `tracker_type: Type[CudaRNGStateTracker]`to `parallelize_module` and change the tracker initialization logic to:
```
if not isinstance(random._rng_tracker, tracker_type):
    random._rng_tracker = tracker_type()
```

wanchaol replied:
Sure, I think this can be done in a follow up PR.

XilunWu replied:
FYI I addressed this in #103910

XilunWu commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -219,29 +225,26 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
-        # if op is a random op, adjust Philox RNG state to maintain synchronization
-        if op_call in random_ops and is_rng_supported_mesh(mesh):
-            set_post_op_offset(dtensor_arg._spec, old_offset)
+            if random._rng_tracker.distribute_region_enabled:
```
moved this check into `distribute_region`

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +59,242 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
-        spec (:class:`DTensorSpec`): the spec of the DTensor object on which
-            we prepare the offset for running random ops.
+        self._states = {}
+        self._devices = [torch.cuda.current_device()]
```
The reason I store `current_device` into a list is, the `fork_rng()` takes in a `List` argument. Then it seems more appropriate to call it `_devices` because it is a list even though it only has one element. WDYT?

XilunWu commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +59,242 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+            offset = (self.rng_states[name])[8:]
+            self.rng_states[name] = torch.cat([seed_tensor, offset])
 
+    def get_offset(self, name: str) -> int:
```
I think they're basic utility functions and probably will be used in other subclasses.

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -16,17 +17,22 @@
 
 
+def _is_random_op(op):
+    aten = torch.ops.aten
+    random_ops = [
+        aten.native_dropout.default,
+        aten.normal_.default,
+        aten.uniform_.default,
+    ]
```
why do we create this list within the function? Any particular reason?

XilunWu replied:
it's simply to help check if an operator is a random op. we can extend this list when we are going to support more random operators.

fduwjj replied:
No, my question is why do you make this list a local variable inside a function?

XilunWu replied:
n00b question on python: if I make it a global variable, will there be a performance gain? Is this the alternative approach you propose?
```
random_ops = [...]

def _is_random_op(op):
    return op in random_ops
```

fduwjj replied:
Well, at least:
```
return op in [
        aten.native_dropout.default,
        aten.normal_.default,
        aten.uniform_.default,
]
```
?

And make aten = torch.ops.aten a const?

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -219,29 +225,21 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
+        if _is_random_op(op_call) and is_rng_supported_mesh(mesh):
+            if not random._rng_tracker:
+                raise RuntimeError(
+                    "A CudaRNGStateTracker instance must be instantiated "
+                    "before executing a random op over a DTensor."
```
Can we also elaborate more on how user can fix this situation? Maybe giving an example like "_rng_tracker = OffsetBasedRNGTracker()"?

XilunWu replied:
Yes! I'll append a guide to this error message.

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: 

-------WIP-------

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
4. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +62,238 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+            (x, y, z) is simply (z, x) by taking(rank_coord[dim_map[0]],rank_coord[dim_map[1]]).
+            Following this calculation,
+            rank 0 and rank 2 holds the shard of coord (0, 0);
+            rank 1 and rank 3 holds the shard of coord (0, 1);
+            rank 4 and rank 6 holds the shard of coord (1, 0);
```
The last placement is Shard(0), which means 0 and 1 are sharded in dim 0, why rank 1 hold the shard coord of (0, 1) not (1, 0)?

XilunWu replied:
`[Shard(1), Replicate(), Shard(0)]` means the dim 0 of DTensor is sharded on mesh dim 2 and DTensor dim 1 is sharded on mesh dim 0. The mesh coordinate of rank 1 is (1, 0, 0). So the shard coordinate owned by rank 1 should be (0, 1). 
cc @wanchaol

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +62,238 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+                    return_offset=False,
+                )[0]
+
+        local_size = math.prod(local_size_on_rank_0)
```
Caveat: `math.prod()` is a new function available in Python versions 3.8 and later..

XilunWu replied:
what is the standard way we resolve this backward compatibility in this case? I think we can:
1. add a function wrapping `math.prod`. If the python version is lower than 3.8, do manual calculation; otherwise calling `math.prod`.
2. write a function using [reduce()](https://docs.python.org/2.7/library/functions.html#reduce) and [operator.mul()](https://docs.python.org/3/library/operator.html#operator.mul)

wanchaol replied:
You can try to use this util function instead https://github.com/pytorch/pytorch/blob/38a29324b0c219ca11bcfe7c8223b4cdc417cffb/torch/distributed/_tensor/ops/utils.py#L71

XilunWu replied:
thx for the link. barely forgot this util function...

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -99,210 +62,238 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+        # get current RNG offset
+        current_offset = self.get_offset("parallel-rng")
+
+        # pytorch: offset must be multiple of 4
+        # source: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp
```
So basically if offset does not have this constraint, we can pretty much mimic the "perfect" random op scenario right?

XilunWu replied:
`offset` forbids the mimic, but not because of the rounding requirement. It's how `offset` is used in pytorch to generate random numbers that forbids our attempt to mimic. :-(

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -11,69 +12,29 @@
-    assert isinstance(
-        device_mesh, DeviceMesh
-    ), f"expect a DeviceMesh but {type(device_mesh)} was passed in."
+def is_rng_supported_mesh(device_mesh: DeviceMesh) -> bool:
```
I feel this is an internal API?

fduwjj commented on torch/distributed/_tensor/random.py:
```diff
@@ -11,69 +12,29 @@
 
 
-def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
+def manual_seed(seed: int, device_mesh: DeviceMesh, tp_dim: int = 0) -> None:
```
ditto.

fduwjj reviewed (commented):
Have some QQs. Nice work!

wanchaol commented:
@XilunWu It seems like the CI failures are correlated, see:

RuntimeError: ERROR: Failed to import modname=torch.testing._internal.distributed._tensor.common_dtensor with modpath=C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\distributed\_tensor\common_dtensor.py
2023-06-23T06:18:05.8232611Z     Caused by: NameError("name 'CudaRNGStateTracker' is not defined")

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: 

-------WIP-------

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
4. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

XilunWu commented:
@wanchaol just fixed the CI issue

wanchaol commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -219,29 +225,23 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
-        # if op is a random op, adjust Philox RNG state to maintain synchronization
-        if op_call in random_ops and is_rng_supported_mesh(mesh):
-            set_post_op_offset(dtensor_arg._spec, old_offset)
+        # For DTensor random operator, run it within a parallel region
```
nit: this comment should be within the if branch

wanchaol commented on torch/distributed/_tensor/random.py:
```diff
@@ -11,69 +12,29 @@
     Args:
         seed (int): The desired seed.
         device_mesh (:class:`DeviceMesh`): The device mesh to set the seed.
+        tp_dim (int, optional): The mesh dimension where to apply Tensor Parallel
```
I don't think this `tp_dim` arg make sense as we haven't use it anywhere inside this function.

I think we should make this arg be sth like `rng_tracker_type` instead, and expose this `manual_seed` API as the API when user want to manually initialize the RNGTracker, and user should have the flexibility to switch between OffsetBasedRNGTracker and the TPRNGTracker.

This might be good as a follow up PR after the TPRNGTracker PR, but I think we should remove this `tp_dim` arg in this PR before merging

wanchaol commented on test/distributed/_tensor/test_random_ops.py:
```diff
@@ -263,6 +245,60 @@ def test_deterministic_uniform_2d(self):
+        # allgather the local tensors
+        dtensor = DTensor.from_local(dtensor.to_local(), device_mesh, [Shard(0)])
+        dtensor = dtensor.redistribute(device_mesh, [Replicate()])
+        local_tensor = dtensor.to_local()
```
hmm maybe we should call `device_mesh.allgather(dtensor.to_local())` directly instead of creating a sharded then converting to replicated? It's kind of weird if the only thing we want to do here is to allgather

wanchaol commented on test/distributed/_tensor/test_random_ops.py:
```diff
@@ -263,6 +245,60 @@ def test_deterministic_uniform_2d(self):
+        dtensor.uniform_()
+
+        dtensor = DTensor.from_local(dtensor.to_local(), device_mesh, [Shard(0)])
+        dtensor = dtensor.redistribute(device_mesh, [Replicate()])
```
ditto: device_mesh.allgather seems more clean

wanchaol reviewed (approved):
Great work! Have a few comments inlined.

XilunWu committed:
Update on "[DTensor][Random] Introduce CudaRNGStateTracker to maintain parallel RNG state for DTensor"


# Change
This PR adds two classes to DTensor:

1. `CudaRNGStateTracker`:  `CudaRNGStateTracker` stores Random Number Generator (RNG) state (a `ByteTensor` object) in a `dict`, mapping from a corresponding tag to each state tensor. It also provides a set of convenient utility methods to help access/modify the state tensors. The most important interface is `_distribute_region` which will be used when DTensor executes a random op (an operator that calls RNG).

2. `OffsetBasedRNGTracker`: 

-------WIP-------

- `_manual_seed`: create an associated RNG state according to the passed in `seed` and store it in `_rng_tracker`. The subclass implementation can decide whether or not to apply this state to the default generator.
- `_parallel_region`: create a context within which the RNG state stored in `_rng_tracker` will be applied before entering 
 the region and restored after exiting.

We also add two subclass of `ParallelRNGStyle`, each representing a different style of parallel random number generation:

1. `DTensorRNG`: our default Parallel RNG style, using the same RNG state across all participating ranks. When DTensor random ops are being executed, random offset will be computed for each rank based on tensor sharding to guarantee that random ops generate same result on `Replicate` part and different results on `Shard` parts.
4. `MegatronRNG`: This style mimics the approach in [Megatron](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/tensor_parallel/random.py) for Tensor Parallel. Briefly speaking, all ranks use the same RNG state for `non-tensor-parallel-region` computation and a unique to `tensor-parallel-rank` RNG state for `tensor-parallel-region`. This means, the computation result of random ops within a Tensor Parallel group will be different, but that will be the same within a Data Parallel group.

The style we use in program can be switched by assigning a new `ParallelRNGStyle` instance to `_parallel_rng_style`.

# Warning

- With `Multi-threaded ProcessGroup`, the global variable `_parallel_rng_style` and `_rng_tracker` will be shared among threads(ranks) and cause issue. We need to figure out a compatible solution for that.

- The RNG state may be asynchronous outside of participating ranks. It is harmless in our current use case of submesh though.

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge
